### ISSUE TITLE
Missing instance of Read for ParseError (or similar functionality).

### ISSUE THREAD

**Issue by expipiplus1** [mrkkrp/megaparsec#98]
It would be useful for testing parsers to be able to read an expected `ParseError` from a file and compare to the result of running a parser. At the moment the only (simple) way to do this is to compare the parse errors as strings, which is a little gross.

It's also a little odd that the instance for `Show` for `ParseError` pretty prints the error, rather than doing what `derive Show` would do. I presume that this is due to historical reasons?

**Comment by mrkkrp** [comment]
We cannot allow construction of `ParseError` in any way other than via smart constructors from `Text.Megaparsec.Error` module. At least for now, because we need to keep `[Messages]` sorted to properly handle comparison of `ParseError`s. So, currently derived `Read` instance would open the possibility to construct invalid `ParseError`s. I'm thinking of using `Set` instead of `[Message]` in Megaparsec 5.0.0 (need to check how this affects performance). If we swich to `Set`, then there will be no way to constuct “invalid” `ParseError` even if we export `ParseError` constructor and derive `Read`. I also like the idea of deriving `Show` and introducing `parseErrorPretty` or similar function for pretty-printing.

**Comment by expipiplus1** [comment]
Did you mean `showErrorPretty` in the last line there. If so hang onto `parseErrorPretty` it'd be a good solution to reading `ParseError`s from strings.

**Comment by mrkkrp** [comment]
`parseErrorPretty` is meant for pretty printing (i.e. it's name of type plus the word “pretty”), but it's not that important how it's called.

----

Although to derive `Read` instance of `ParseError` we need to derive `Read` of `SourcePos` too, that may be tricky, since line and column numbers must be greater than 1 — again we should only allow construction the `SourcePos` via smart constructors because of this.

Derived `Show` instances for these things will probably be in Megaparsec 5.0.0, but I'm not sure about `Read` instances.

**Comment by expipiplus1** [comment]
In case anyone runs into the same issue, here's a set of `newtype` wrappers to read and show `ParseError` and friends.

``` haskell
import           GHC.Read                        (Read (..), expectP)
import qualified Text.Megaparsec.Error           as M
import qualified Text.Megaparsec.Pos             as M
import           Text.ParserCombinators.ReadPrec (choice, prec, step)
import           Text.Read                       (parens)
import           Text.Read.Lex                   (Lexeme (Ident))

--------------------------------------------------------------------------------
-- Workaround for https://github.com/mrkkrp/megaparsec/issues/98
--------------------------------------------------------------------------------

newtype ShowParseError = ShowParseError{ unShowParseError :: M.ParseError }
  deriving(Eq)

newtype ShowMessage = ShowMessage{ unShowMessage :: M.Message }
  deriving(Eq, Ord)

newtype ShowSourcePos = ShowSourcePos{ unShowSourcePos ::  M.SourcePos }
  deriving(Eq, Ord)

instance Show ShowParseError where
  showsPrec p (ShowParseError e) =
    let ms  = ShowMessage <$> M.errorMessages e
        pos = ShowSourcePos $ M.errorPos e
    in showParen (p >= 11) ( showString "newErrorMessages "
                           . showsPrec 11 ms
                           . showString " "
                           . showsPrec 11 pos
                           )

instance Read ShowParseError where
  readPrec = parens ( prec 10 (do
                        expectP (Ident "newErrorMessages")
                        ms <- map unShowMessage <$> step readPrec
                        pos <- unShowSourcePos <$> step readPrec
                        pure $ ShowParseError (M.newErrorMessages ms pos)
                        )
                    )

instance Show ShowMessage where
  showsPrec p (ShowMessage m) =
    case m of
      M.Unexpected s -> showParen (p >= 11) ( showString "Unexpected "
                                            . showsPrec 11 s
                                            )
      M.Expected s -> showParen (p >= 11) ( showString "Expected "
                                          . showsPrec 11 s
                                          )
      M.Message s -> showParen (p >= 11) ( showString "Message "
                                         . showsPrec 11 s
                                         )

instance Read ShowMessage where
  readPrec = parens ( choice [ prec 10 (do
                                 expectP (Ident "Unexpected")
                                 m <- step readPrec
                                 pure (ShowMessage $ M.Unexpected m)
                                 )
                             , prec 10 (do
                                 expectP (Ident "Expected")
                                 m <- step readPrec
                                 pure (ShowMessage $ M.Expected m)
                                 )
                             , prec 10 (do
                                 expectP (Ident "Message")
                                 m <- step readPrec
                                 pure (ShowMessage $ M.Message m)
                                 )
                             ]
                    )


instance Show ShowSourcePos where
  showsPrec p (ShowSourcePos pos) =
    let name = M.sourceName pos
        line = M.sourceLine pos
        col  = M.sourceColumn pos
    in showParen (p >= 11) ( showString "newPos "
                           . showsPrec 11 name
                           . showString " "
                           . showsPrec 11 line
                           . showString " "
                           . showsPrec 11 col
                           )

instance Read ShowSourcePos where
  readPrec = parens ( prec 10 (do
                        expectP (Ident "newPos")
                        name <- step readPrec
                        line <- step readPrec
                        col <- step readPrec
                        pure $ ShowSourcePos (M.newPos name line col)
                        )
                    )
```

**Comment by mrkkrp** [comment]
That's quite a bit of work. I will try to provide something out-of-box when I get to re-organisation of this part of the library. Thanks for reporting the issue.

**Comment by mrkkrp** [comment]
The desired instances are derived as part of #99 and will be available in Megaparsec 5.0.0. Closing, and thank you for opening the issue.

**Comment by expipiplus1** [comment]
Thank you!
